internal/engine/k8swatch/pod_watch.go

```diff
@@ -37,79 +65,196 @@ type PodWatcher struct {
+//
+// This is intended for use in tests.
+func (w *PodWatcher) ExtraSelectors(key types.NamespacedName) []labels.Selector {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	var ret []labels.Selector
+	for _, s := range w.extraSelectors[key] {
+		ret = append(ret, s.DeepCopySelector())
+	}
+	return ret
+}
 
+// Claimant returns the key that has uniquely claimed a particular K8s entity UID to prevent
+// duplicate events.
+//
+// This is intended for use in tests.
+func (w *PodWatcher) Claimant(uid types.UID) types.NamespacedName {
 	w.mu.Lock()
 	defer w.mu.Unlock()
+	return w.claimedUIDs[uid]
+}
+
+// reconcile manages namespace watches for the modified KubernetesDiscovery object.
+func (w *PodWatcher) reconcile(ctx context.Context, st store.RStore, key types.NamespacedName) {
+	state := st.RLockState()
+	kd := state.KubernetesDiscoveries[key]
+	st.RUnlockState()
+
+	if kd == nil {
+		// this object doesn't exist in store (i.e. it was deleted) so clean up
+		// any watches + other metadata
+		delete(w.kubernetesDiscoveryManifest, key)
+		delete(w.extraSelectors, key)
+		for _, watcher := range w.watchedNamespaces {
+			// remove this key from the list of watchers; if it was the last watcher, the
+			// namespace watch will be cleaned up afterwards
+			delete(watcher.watchers, key)
+		}
+		for uid, k := range w.claimedUIDs {
+			if k == key {
+				delete(w.claimedUIDs, uid)
+			}
+		}
+		return
+	}
 
-	w.extraSelectors = taskList.extraSelectors
+	if manifestName, ok := kd.Annotations[v1alpha1.AnnotationManifest]; ok && manifestName != "" {
+		// PodChangeAction is currently tied to manifest so keep a mapping of API objects to manifests
+		// (this will be removed shortly once PodChangeAction -> KubernetesDiscoveryUpdateStatusAction)
+		w.kubernetesDiscoveryManifest[key] = model.ManifestName(manifestName)
+	}
 
-	for _, teardown := range taskList.teardownNamespaces {
-		watcher, ok := w.watcherKnownState.namespaceWatches[teardown]
-		if ok {
-			watcher.cancel()
+	w.extraSelectors[key] = nil
+	for _, labelValues := range kd.Spec.ExtraSelectors {
+		ls := selectorFromLabels(labelValues)
+		if !ls.Empty() {
+			w.extraSelectors[key] = append(w.extraSelectors[key], ls)
 		}
-		delete(w.watcherKnownState.namespaceWatches, teardown)
 	}
 
-	for _, setup := range taskList.setupNamespaces {
-		w.setupWatch(ctx, st, setup)
+	seenUIDs := make(map[types.UID]bool)
+	seenNamespaces := make(map[k8s.Namespace]bool)
+	for _, toWatch := range kd.Spec.Watches {
+		namespace := k8s.Namespace(toWatch.Namespace)
+		if !seenNamespaces[namespace] {
+			w.setupWatch(ctx, st, namespace, key)
+			seenNamespaces[namespace] = true
+		}
+
+		id := types.UID(toWatch.UID)
+		if id == "" || seenUIDs[id] {
+			continue
+		}
+		seenUIDs[id] = true
+
+		if claimant, ok := w.claimedUIDs[id]; !ok || claimant != key {
+			// either this UID has not been claimed (so it's new to Tilt)
+			// or the claim has changed (so it's new to this KubernetesDiscovery object)
```

Comment:
i think this is going to lead to infinite loops where claims bounce back and forth between objects? (this is what the old code was protecting against)

i wonder if the data model should simply allow multiple discovery objects to have the same UID, which would make this more consistent with how it works where you set up nsWatch and clean up abandoned ones

Reply:
RE: data model, absolutely - that was my intent, in subsequent changes I pushed the "claim" logic up to `k8swatch.ManifestSubscriber` so it handles not creating "duplicates" from its perspective, but the actual watch logic doesn't have to care.

I'd originally chunked it to limit scope of this, but yeahhh this is definitely wrong, so gonna pull those changes in now because it's so much easier to reason about that way

Reply:
Moved the logic around in d2f14cc and put in a test case around a bunch of behaviors here. There's some lingering weirdness (that's heavily documented) in `PodWatcher` for the extra label selector cases.